src/cargo/sources/git/utils.rs

```diff
@@ -207,56 +207,54 @@ impl GitDatabase {
         self.repo.revparse_single(&oid.to_string()).is_ok()
     }
 
-    /// [`GitReference::resolve`]s this reference with this database.
+    /// [`resolve_ref`]s this reference with this database.
     pub fn resolve(&self, r: &GitReference) -> CargoResult<git2::Oid> {
-        r.resolve(&self.repo)
+        resolve_ref(r, &self.repo)
     }
 }
 
-impl GitReference {
-    /// Resolves self to an object ID with objects the `repo` currently has.
-    pub fn resolve(&self, repo: &git2::Repository) -> CargoResult<git2::Oid> {
-        let id = match self {
-            // Note that we resolve the named tag here in sync with where it's
-            // fetched into via `fetch` below.
-            GitReference::Tag(s) => (|| -> CargoResult<git2::Oid> {
-                let refname = format!("refs/remotes/origin/tags/{}", s);
-                let id = repo.refname_to_id(&refname)?;
-                let obj = repo.find_object(id, None)?;
-                let obj = obj.peel(ObjectType::Commit)?;
-                Ok(obj.id())
-            })()
-            .with_context(|| format!("failed to find tag `{}`", s))?,
-
-            // Resolve the remote name since that's all we're configuring in
-            // `fetch` below.
-            GitReference::Branch(s) => {
-                let name = format!("origin/{}", s);
-                let b = repo
-                    .find_branch(&name, git2::BranchType::Remote)
-                    .with_context(|| format!("failed to find branch `{}`", s))?;
-                b.get()
-                    .target()
-                    .ok_or_else(|| anyhow::format_err!("branch `{}` did not have a target", s))?
-            }
+/// Resolves self to an object ID with objects the `repo` currently has.
```

Comment:
```suggestion
/// Resolves an [`GitReference`] to an object ID with objects the `repo` currently has.
```

Though personally I prefer an extension trait for grouping these methods.

Reply:
For `PackageIdSpec`, the type is widely used and so having a single trait to pull in and access the functionality seems reasonable.  For this, its pretty limited in where its used (with one exception) which makes me lean the other way.  Or to put this another way, if I were writing this code from scratch, this is likely how I would have done it.